R/PipeOpLearnerCV.R

```diff
@@ -159,53 +165,71 @@ PipeOpLearnerCV = R6Class("PipeOpLearnerCV",
       } else {
         multiplicity_recurse(self$state, clone_with_state, learner = private$.learner)
       }
+    },
+    resampling = function(val) {
+      if (!missing(val)) {
+        if (!identical(val, private$.resampling)) {
+          stop("$resampling is read-only.")
+        }
+      }
+      private$.resampling
     }
   ),
   private = list(
     .train_task = function(task) {
       on.exit({private$.learner$state = NULL})
-
-      # Train a learner for predicting
+      # train a learner for predicting
       self$state = private$.learner$train(task)$state
-      pv = private$.crossval_param_set$values
 
-      # Compute CV Predictions
-      if (pv$method != "insample") {
-        rdesc = mlr_resamplings$get(pv$method)
-        if (pv$method == "cv") rdesc$param_set$values = list(folds = pv$folds)
-        rr = resample(task, private$.learner, rdesc)
-        prds = as.data.table(rr$prediction(predict_sets = "test"))
-      } else {
-        prds = as.data.table(private$.learner$predict(task))
-      }
+      # compute resampled predictions
+      rr = resample(task, private$.learner, private$.resampling)
+      prds = as.data.table(rr$prediction(predict_sets = "test"))
 
-      private$pred_to_task(prds, task)
+      private$.pred_to_task(prds, task)
     },
 
     .predict_task = function(task) {
       on.exit({private$.learner$state = NULL})
       private$.learner$state = self$state
-      prediction = as.data.table(private$.learner$predict(task))
-      private$pred_to_task(prediction, task)
+      prds = as.data.table(private$.learner$predict(task))
+      private$.pred_to_task(prds, task)
     },
 
-    pred_to_task = function(prds, task) {
-      if (!is.null(prds$truth)) prds[, truth := NULL]
-      if (!self$param_set$values$resampling.keep_response && self$learner$predict_type == "prob") {
+    .pred_to_task = function(prds, task) {
+      if (!self$param_set$values$keep_response && self$learner$predict_type == "prob") {
         prds[, response := NULL]
       }
-      renaming = setdiff(colnames(prds), c("row_id", "row_ids"))
-      setnames(prds, renaming, sprintf("%s.%s", self$id, renaming))
+      renaming = setdiff(colnames(prds), c("row_ids", "truth"))
+      setnames(prds, old = renaming, new = sprintf("%s.%s", self$id, renaming))
+      setnames(prds, old = "truth", new = task$target_names)
+      row_reference = paste0("pre.", self$id)
+      while (row_reference %in% task$col_info$id) {
+        row_reference = paste0(row_reference, ".")
```

Comment:
instead throw an error when IDs collide, user has to change PipOp's ID then. In any case IDs are unique within graph so usually shouldn't be a problem.